### PR TITLE
Ideal geometry2015 muon

### PR DESCRIPTION
**Pull request by pietverwilligen**
This Pull Request creates two geometries:
- Ideal2015Muon
- Extended2015Muon
  both contain all the Phase-2 Muon upgrades: ME0, GE1/1, GE2/1, RE3/1, RE4/1 on top of 
  the Run-2 geometry, and without ECAL and HCAL.  

This PR allows the integration of the Muon Upgrade code in 75X, because
now there is a geometry in which ME0 is included.

While testing the Geometry I noticed that the ME0 SimHits were not produced, therefore also a modification to the OscarProducer was necessary. ME0 Pseudo-Digis are not yet produced / kept, but that will be for another PR.

Please test and sign.
Thanks
Piet

### CODE REVIEW

**Comment by pietverwilligen** [comment]
RecoLocalMuon/GEMRecHit is involved because I have my test files there ... apologies Slava. I did not touch any source code there.
greets
Piet

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
Piet,
Is this a solution to get things to work for muons now (put the Phase-II muon geometry on top of otherwise Run-2/2015 detector)?

**Comment by pietverwilligen** [comment]
Yes, it is exactly that. I had to remove the calorimeters though.

**Comment by dildick** [comment]
@pietverwilligen About the ME0 pseudo-digis. Indeed, they are not being produced by default. That is why we have a bunch of functions that (https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/SimMuon/GEMDigitizer/python/customizeGEMDigi.py) can be used in a cmsDriver command. These functions basically do the same as https://github.com/cms-sw/cmssw/blob/CMSSW_6_2_X_SLHC/SLHCUpgradeSimulations/Configuration/python/me0Customs.py#L24. Such an me0Customs file is not to be included in CMSSW_7_5_X though. Instead we have to customize the process using "era" type functions. However, there is no era for 2019 or 2023 upgrade studies yet.

File: Geometry/CMSCommonData/python/cmsIdealGeometry2015XML_cfi.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
         'Geometry/CMSCommonData/data/cmsBeam.xml',
         'Geometry/CMSCommonData/data/muonMB.xml',
         'Geometry/CMSCommonData/data/muonMagnet.xml',
+        'Geometry/TrackerCommonData/data/trackerParameters.xml',
```
Comment (ianna) [inline Geometry/CMSCommonData/python/cmsIdealGeometry2015XML_cfi.py]: @pietverwilligen - this is part of a separate PR. I think, it is better to remove this file from here.

File: SimG4Core/Application/plugins/OscarProducer.cc [hunk 2]
```diff
@@ -102,6 +102,8 @@ OscarProducer::OscarProducer(edm::ParameterSet const & p)
   produces<edm::PSimHitContainer>("MuonCSCHits");
   produces<edm::PSimHitContainer>("MuonRPCHits");
   produces<edm::PSimHitContainer>("MuonGEMHits");
+  produces<edm::PSimHitContainer>("MuonME0Hits");
+
```
Comment (makortel) [inline SimG4Core/Application/plugins/OscarProducer.cc]: Please make the same change to OscarMTProducer too (that is already the default in 75X).

**Comment by pietverwilligen** [comment]
Hi All

I have updated the pull request, can you please test?

@dildick thanks for pointing out, I had no time Thursday to look into this, but now I have included it. this pull requests has everything to test the ME0 chain working up to the production of digis.

@ianna i removed the two files from the pull request

@makortel i apllied the change also in OscarMTProducer.cc

greets
Piet

**Comment by ianna** [comment]
+1

**Comment by pietverwilligen** [comment]
It seemed to have failed in the first step of the 4.53 2012B Photon test...
Its not really clear to me why, because that is just the das query, right?

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8995/22/runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT+MINIAODDreHLT/

**Comment by slava77** [comment]
@cmsbuild what happened here with the tests triggering?
there is a +1 from jenkins at https://github.com/cms-sw/cmssw/pull/8995#issuecomment-100939983 for the same commit and then shortly after another test was done with a -1 https://github.com/cms-sw/cmssw/pull/8995#issuecomment-100980314

**Comment by nclopezo** [comment]
+1
Hi, 
They both were done with the same IB, the second round of tests had a DAS issue. We have had some issues with the jenkins server, and after the server is restarted we restart some tests that could have not finished. This is why there are 2 tests one after another. I wrote the "+1" to mark this PR as `tests-approved`.

**Comment by civanch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/8995	514a1c1
- reco parts are minimally affected (only configs in RecoLocalMuon/GEMRecHit/test)
- jenkins tests complete ok and show no differences in comparisons, as expected

**Comment by davidlange6** [comment]
@ianna - is the sim group on board with managing these unphysical geometries in this way? I'm  worried both about the "2015" in the name and wondering if we would be better of putting these in a test directory somewhere instead of the official place to make it more difficult for non-experts to get the wrong answer

**Comment by civanch** [comment]
I agree that name/directory for upgrade geometry should be clear. In this PR we do not have a direct clash but a potential  problem in long term. Probably, this PR may be merged but we should re-discuss when we discuss muon upgrade software (likely next week).

**Comment by pietverwilligen** [comment]
Hi Matti

When doing some checks I see that the ME0SimHits are produced with
DetUnitId = 0.
Can you point me to the piece of code where they are exactly produced, such
that I can investigate it?


Hi Vladimir, David, I agree that the name of those technical geometries is
a bit unfortunate ...

greets
Piet

On Tue, May 12, 2015 at 3:26 PM, Vladimir Ivantchenko <
notifications@github.com> wrote:

> I agree that name/directory for upgrade geometry should be clear. In this
> PR we do not have a direct clash but a potential problem in long term.
> Probably, this PR may be merged but we should re-discuss when we discuss
> muon upgrade software (likely next week).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/8995#issuecomment-101276001>.
>



-- 
Piet Verwilligen

INFN -- Sezione di Bari
Via E. Orabona 4
I-70125 Bari, Italy
Phone: +39 345 74 70 642

**Comment by makortel** [comment]
Piet, I don't really know, but my guess would somewhere in
https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/SimG4CMS/Muon/interface/MuonSensitiveDetector.h
https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/SimG4CMS/Muon/interface/MuonSlaveSD.h

Vladimir could know better.

**Comment by pietverwilligen** [comment]
Hi Vladimir

When doing some checks I see that the ME0SimHits are produced with
DetUnitId = 0.
Can you point me to the piece of code where they are exactly produced, such
that I can investigate it?
Thanks
Piet

On Tue, May 19, 2015 at 9:27 AM, Matti Kortelainen <notifications@github.com
> wrote:

> Piet, I don't really know, but my guess would somewhere in
>
> https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/SimG4CMS/Muon/interface/MuonSensitiveDetector.h
>
> https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/SimG4CMS/Muon/interface/MuonSlaveSD.h
>
> Vladimir could know better.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/8995#issuecomment-103379797>.
>



-- 
Piet Verwilligen

INFN -- Sezione di Bari
Via E. Orabona 4
I-70125 Bari, Italy
Phone: +39 345 74 70 642

**Comment by ianna** [comment]
@davidlange6 - Extended2015 scenario is the base one for this, hence it is reflected in the name. In any case, the new scenario supposed to be very short-lived and it will not affect the base line. I've already started 2023 line which will replace this new scenario. Something like "Test" could be added to the name, but it has to be in a current place for cmsDriver to pick it up without too much hassle. Non-experts should use DB geometry and I'm sure, one would be making an extra effort to pick this one up.

**Comment by davidlange6** [comment]
@ianna, @pietverwilligen - indeed something like Test, Dev, and maybe GEM too is needed to make it really obvious that this is not a 2015 geometry that could be of interest.

**Comment by pietverwilligen** [comment]
I have changed the names to 2015Muon_GEMDev.
I have removed the IdealGeometry2015Muon, since we do not need 2 geometries to do the test ... 
so keeping only GeometryExtended2015Muon that has been renamed to GeometryExtended2015MuonXML_GEMDev

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test
new IB CMSSW_7_5_X_2015-05-23-1100 is in and seems ok at least on the build side

**Comment by civanch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/8995	6431479
- reco is minimally affected (only files in /test)
- jenkins tests completed and show no relevant differences in comparisons with the baseline 
    - IsoTkMu HLT plots in DQM have occasional small differences, which persists in IB tests for the past few days already

**Comment by alja** [comment]
+1

**Comment by davidlange6** [comment]
@pietverwilligen - thanks - I worry that someone is assuming that '_' are not part of the geometry name. Could you just remove the _ and I'll merge the PR

**Comment by pietverwilligen** [comment]
done, please test again.

**Comment by pietverwilligen** [comment]
Hi Matti

Thanks

I found that simhits are produced exactly in
https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/SimG4CMS/Muon/src/MuonSensitiveDetector.cc

Thanks, very useful
greets
Piet

On Tue, May 19, 2015 at 9:27 AM, Matti Kortelainen <notifications@github.com
> wrote:

> Piet, I don't really know, but my guess would somewhere in
>
> https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/SimG4CMS/Muon/interface/MuonSensitiveDetector.h
>
> https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/SimG4CMS/Muon/interface/MuonSlaveSD.h
>
> Vladimir could know better.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/8995#issuecomment-103379797>.
>



-- 
Piet Verwilligen

INFN -- Sezione di Bari
Via E. Orabona 4
I-70125 Bari, Italy
Phone: +39 345 74 70 642

**Comment by ianna** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/8995	a461e51
- reco is minimally affected (only config files in /test)
- jenkins tests pass and comparisons with the baseline show no difference

**Pull request merged**